SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosStatusUpdateHandler.java

```diff
@@ -346,14 +347,57 @@ private StatusUpdateResult unsafeProcessStatusUpdate(
           taskIdObj
         );
         return StatusUpdateResult.KILL_TASK;
-      } else if (
-        maybeTaskHistory.isPresent() &&
-        status.getReason() == Reason.REASON_AGENT_REREGISTERED
-      ) {
+      } else if (status.getReason() == Reason.REASON_AGENT_REREGISTERED) {
         boolean lbRemovalStarted = taskManager
           .getLoadBalancerState(taskIdObj, LoadBalancerRequestType.REMOVE)
           .isPresent();
         if (lbRemovalStarted) {
+          LOG.info(
+            "LB removal for recovered task {} was already started. Attempting to clear and start as new task",
+            taskId
+          );
+          if (
+            taskManager.reactivateTask(
+              taskIdObj,
+              taskState,
+              newTaskStatusHolder,
+              Optional.ofNullable(status.getMessage()),
+              status.hasReason()
+                ? Optional.of(status.getReason().name())
+                : Optional.empty()
+            )
+          ) {
+            Optional<SingularityTask> maybeTask = taskManager.getTask(taskIdObj);
+            Optional<SingularityRequestWithState> maybeRequest = requestManager.getRequest(
+              taskIdObj.getRequestId()
+            );
+            if (
+              maybeTask.isPresent() &&
+              maybeRequest.isPresent() &&
+              maybeRequest.get().getState().isRunnable()
+            ) {
+              LOG.info(
+                "Task {} can be recovered. Clearing LB state and enqueuing check as new task",
+                taskId
+              );
+              taskManager.clearLoadBalancerHistory(taskIdObj);
```

Comment:
Is there any chance we get into a weird state if we don't clear this before attempting to reactivate the task? I'm thinking new LB stuff put into zk between the two calls?

Reply:
My initial thought is that, if we have gotten here, we already processed a task_lost update for the task and considered it terminal. However, there is the possibility that a removal might still be in progress if the removal was slow 🤔 , good point

Reply:
Added handling for possible in-progress LB removal now ✅